tests: Add processor switching tests

erin-le opened the pull request:
This PR adds tests for processor switching between combinations of atomic, timing, minor, and o3 cpus on the x86,
Arm, and RISC-V ISAs. The SE mode tests use the matrix multiplication workloads and run tests on one core, and the FS mode tests use the NPB CG size S workload and run tests on 4, 8, and 16 cores.

erin-le committed:
tests: Add cross product processor switching tests

This commit adds tests for processor switching between
combinations of atomic, timing, minor, and o3 cpus on the x86,
arm, and riscv ISAs. The SE mode tests use the matrix
multiplication workloads and run tests on one core, and
the FS mode tests use the NPB CG size S workload and run tests
on 4, 8, and 16 cores.

erin-le committed:
tests: change local filepaths to obtain resource calls

erin-le committed:
tests: temporarily remove riscv ubuntu boot tests

This commit temporarily removes the riscv ubuntu boot tests, as
the checkpoints needed for them are not yet available.

erin-le committed:
tests: improve string formatting

erin-le committed:
tests: add scripts for processor switch test checkpoints

This commit adds the scripts needed to take checkpoints used by
the full system processor switch tests.

erin-le committed:
tests: add readme to processor switching tests

erin-le committed:
tests: modify file names for processor switching tests

erin-le committed:
tests: minor change to processor switch tests

erin-le committed:
tests: remove extra imports and comments from proc switch tests

Harshil2107 commented on tests/gem5/processor_switch_tests/configs/checkpoint/cross-product-switch-afterboot-take-cpt-arm.py:
```diff
@@ -0,0 +1,136 @@
+        "arm64-bootloader-foundation", resource_version="1.0.0"
+    ),
+    readfile_contents="/home/gem5/NPB3.4-OMP/bin/cg.S.x; sleep 5;",
+)
```
```suggestion
board.set_workload(
   obtain_resource(
        "arm-ubuntu-24.04-npb-cg-s", resource_version="2.0.0"
    )
)
```

Harshil2107 commented on tests/gem5/processor_switch_tests/configs/checkpoint/cross-product-switch-afterboot-take-cpt-riscv.py:
```diff
@@ -0,0 +1,116 @@
+        "riscv-bootloader-opensbi-1.3.1", resource_version="1.0.0"
+    ),
+    readfile_contents="/home/gem5/NPB3.4-OMP/bin/cg.S.x; sleep 5;",
+)
```
```suggestion
board.set_workload(
    obtain_resource(
        "riscv-ubuntu-24.04-npb-cg-s", resource_version="1.0.0"
    )
)
```

Harshil2107 commented on tests/gem5/processor_switch_tests/configs/checkpoint/cross-product-switch-afterboot-take-cpt-x86.py:
```diff
@@ -0,0 +1,115 @@
+        "root=/dev/sda2",
+    ],
+    readfile_contents=f"/home/gem5/NPB3.4-OMP/bin/cg.S.x; sleep 5;",
+)
```
```suggestion
board.set_workload(
   obtain_resource(
        "x86-ubuntu-24.04-npb-cg-s", resource_version="3.0.0"
    )
)
```

Harshil2107 reviewed (changes requested):
The changes looks good to me. A few small things, I think we should directly call the predefined workload when taking the checkpoint instead of setting it up via `set_kernel_disk_workload`.

 Also should these checkpoints be uploaded to resources?

erin-le committed:
tests: update checkpoint taking scripts to use NPB workloads

This commit updates the scripts that take checkpoints for the
processor switching tests to use the NPB workloads instead of
setting disk images, kernels, etc. individually using
set_kernel_disk_workload. The bootloader resource for the Arm boot
in `cross-product-switch-afterboot.py` has also been updated to
match this change.

erin-le committed:
tests: add more info to the README of the processor switch tests

This commit adds more information about generating checkpoints
using the provided scripts to the README of the processor switch
tests.
